src/Evolution/Systems/NewtonianEuler/Limiters/Flattener.cpp

```diff
@@ -0,0 +1,152 @@
+#include "NumericalAlgorithms/Spectral/Mesh.hpp"
+#include "PointwiseFunctions/Hydro/EquationsOfState/EquationOfState.hpp"
+#include "Utilities/ConstantExpressions.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/MakeArray.hpp"
+
+namespace NewtonianEuler::Limiters {
+
+template <size_t VolumeDim, size_t ThermodynamicDim>
+size_t flatten_solution(
+    const gsl::not_null<Scalar<DataVector>*> mass_density_cons,
+    const gsl::not_null<tnsr::I<DataVector, VolumeDim>*> momentum_density,
+    const gsl::not_null<Scalar<DataVector>*> energy_density,
+    const Mesh<VolumeDim>& mesh,
+    const Scalar<DataVector>& det_logical_to_inertial_jacobian,
+    const EquationsOfState::EquationOfState<false, ThermodynamicDim>&
+        equation_of_state) noexcept {
+  // Note on the design behind the handling of the cell-averaged fields:
+  //
+  // The cell averages are needed for
+  // - sanity-checking the inputs for validity
+  // - flattening in case of negative density
+  // - flattening in case of negative pressure
+  //
+  // In addition, we have the following optimization goals:
+  // - cell averages are only computed once
+  // - cell averages are only computed when needed
+  //
+  // To meet these goals, we create a temporary variable for each field's cell
+  // average. These temporaries live on the stack and should have minimal cost.
+  double mean_density = std::numeric_limits<double>::signaling_NaN();
+  auto mean_momentum =
+      make_array<VolumeDim>(std::numeric_limits<double>::signaling_NaN());
+  double mean_energy = std::numeric_limits<double>::signaling_NaN();
+
+  const auto compute_means = [&mean_density, &mean_momentum, &mean_energy,
+                              &mass_density_cons, &momentum_density,
+                              &energy_density, &mesh,
+                              &det_logical_to_inertial_jacobian]() noexcept {
+    // compute the means
+    const double volume_of_cell =
+        definite_integral(get(det_logical_to_inertial_jacobian), mesh);
+    const auto jacobian_weighted_mean =
+        [&mesh, &det_logical_to_inertial_jacobian,
+         &volume_of_cell](const DataVector& u) noexcept {
+          return definite_integral(get(det_logical_to_inertial_jacobian) * u,
+                                   mesh) /
+                 volume_of_cell;
+        };
+    mean_density = jacobian_weighted_mean(get(*mass_density_cons));
+    for (size_t i = 0; i < VolumeDim; ++i) {
+      gsl::at(mean_momentum, i) =
+          jacobian_weighted_mean(momentum_density->get(i));
+    }
+    mean_energy = jacobian_weighted_mean(get(*energy_density));
+
+    // sanity check the means
+    ASSERT(mean_density > 0., "Invalid mass density input to flattener");
+    if constexpr (ThermodynamicDim == 2) {
+      ASSERT(mean_energy > 0., "Invalid energy density input to flattener");
+    }
+  };
+
+  size_t flattener_status = 0;
```

Comment:
It seems this could be easier to read if it were an enum. It's not at all clear what the number returned means. I'd be happier with either a bool "yes I limited" or an enum "I limited this". If you need the return type to be convertible to a bool, I'd suggest an enum with explicitly specified values